### PR TITLE
LibWeb: Implement `AbortSignal.timeout()`

### PR DESCRIPTION
**Pull request by tcl3**
This method returns a signal that will automatically abort after a given number of milliseconds.

Progress towards #13355

### CODE REVIEW

File: Userland/Libraries/LibWeb/DOM/AbortSignal.cpp [hunk 1]
```diff
@@ -132,4 +134,30 @@ WebIDL::ExceptionOr<JS::NonnullGCPtr<AbortSignal>> AbortSignal::abort(JS::VM& vm
...
+{
+    auto& realm = *vm.current_realm();
+
+    // 1. Let signal be a new AbortSignal object.
+    auto signal = TRY(construct_impl(realm));
+
+    // 2. Let global be signal’s relevant global object.
+    auto& global = realm.global_object();
+    auto* window_or_worker = dynamic_cast<HTML::WindowOrWorkerGlobalScopeMixin*>(&global);
+    VERIFY(window_or_worker);
```
Comment (trflynn89) [inline Userland/Libraries/LibWeb/DOM/AbortSignal.cpp:148]:
```c++
auto& global = HTML::relevant_global_object(*this);
```
:)

(plus whatever casting you have to do to get to `WindowOrWorkerGlobalScopeMixin`).

Comment (trflynn89) [reply Userland/Libraries/LibWeb/DOM/AbortSignal.cpp:148]: FYI to find these AOs, I just click the link from the spec text ("relevant global object" in this case) and grep the code base for the anchor (concept-relevant-global):
```bash
» cgr concept-relevant-global
Userland/Libraries/LibWeb/HTML/Scripting/Environments.cpp
415:// https://html.spec.whatwg.org/multipage/webappapis.html#concept-relevant-global
```

so then I know we can just call that. (Or if the AO hasn't been implemented yet, try to implement it to help our future selves)

File: Userland/Libraries/LibWeb/HTML/WindowOrWorkerGlobalScope.cpp [hunk 2]
```diff
@@ -309,9 +310,7 @@ i32 WindowOrWorkerGlobalScopeMixin::run_timer_initialization_steps(TimerHandler
     };
 
     // 12. Run steps after a timeout given global, "setTimeout/setInterval", timeout, completionStep, and id.
-    auto timer = Timer::create(this_impl(), timeout, move(completion_step), id);
-    m_timers.set(id, timer);
-    timer->start();
+    run_steps_after_a_timeout_impl(timeout, move(completion_step), id);
```
Comment (trflynn89) [inline Userland/Libraries/LibWeb/HTML/WindowOrWorkerGlobalScope.cpp:313]: This bit really belongs in its own commit

File: Userland/Libraries/LibWeb/HTML/WindowOrWorkerGlobalScope.cpp [hunk 3]
```diff
@@ -554,4 +553,18 @@ void WindowOrWorkerGlobalScopeMixin::queue_the_performance_observer_task()
...
+    auto id = m_timer_id_allocator.allocate();
+    return run_steps_after_a_timeout_impl(timeout, move(completion_step), id);
+}
+
+void WindowOrWorkerGlobalScopeMixin::run_steps_after_a_timeout_impl(i32 timeout, Function<void()> completion_step, i32 id)
+{
+    auto timer = Timer::create(this_impl(), timeout, move(completion_step), id);
+    m_timers.set(id, timer);
+    timer->start();
+}
```
Comment (trflynn89) [inline Userland/Libraries/LibWeb/HTML/WindowOrWorkerGlobalScope.cpp:568]: Given that the spec text is:
```c++
and an optional value timerKey:
```

I think we need to implement this as one function, and make the timer key an Optional:
```c++
// header
void run_steps_after_a_timeout_impl(i32 timeout, Function<void()> completion_step, Optional<i32> timer_key = {});

// impl
// https://html.spec.whatwg.org/multipage/timers-and-user-prompts.html#run-steps-after-a-timeout
void WindowOrWorkerGlobalScopeMixin::run_steps_after_a_timeout_impl(i32 timeout, Function<void()> completion_step, Optional<i32> timer_key)
{
    // 2. If timerKey is not given, then set it to a new unique non-numeric value.
    if (!timer_key.has_value())
        timer_key = m_timer_id_allocator.allocate();
}
```

File: Userland/Libraries/LibWeb/HTML/WindowOrWorkerGlobalScope.cpp (hunk 2 above)
Comment (tcl3) [reply Userland/Libraries/LibWeb/HTML/WindowOrWorkerGlobalScope.cpp:313]: I've split this into it's own commit and added (mostly FIXME) spec comments to the function.

File: Userland/Libraries/LibWeb/HTML/WindowOrWorkerGlobalScope.cpp (hunk 3 above)
Comment (tcl3) [reply Userland/Libraries/LibWeb/HTML/WindowOrWorkerGlobalScope.cpp]: I've kept this as 2 functions in order to enforce step 1:
> 1. Assert: if timerKey is given, then the caller of this algorithm is the timer initialization steps. (Other specifications must not pass timerKey.)

I've refactored it to use an Optional as suggested. It seems a bit unnecessary in this case, but it does more closely follow the spec.
